fix: Allow valueFrom in dag arguments parameters. Fixes #11900

nice-pink opened the pull request:
<!-- markdownlint-disable MD041 -->

<!--

### Before you open your PR

- Run `make pre-commit -B` to fix codegen and lint problems (build will fail).
- [Signed-off your commits](https://github.com/apps/dco/) (otherwise the DCO check will fail).
- Used [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/).

### When you open your PR

- PR title format should also conform to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
- Create the PR as draft.
- Once builds are green, mark your PR "Ready for review".

When changes are requested, please address them and then dismiss the review to get it reviewed again.

-->

<!-- Does this PR fix an issue -->

Fixes #11900 

### Motivation

Allow assigning values from config maps to dag.tasks.arguments.parameters.

### Modifications

The validation for dag task arguments was updated to allow `valueFrom`, not only `value`.

### Verification

<!-- TODO: Say how you tested your changes. -->

terrytangyuan commented on workflow/validate/validate_dag_test.go:
```diff
@@ -1039,42 +1039,3 @@ func TestDAGOutputsResolveTaskAggregatedOutputs(t *testing.T) {
 	assert.NoError(t, err)
 }
-
-var dagMissingParamValueInTask = `
```
Maybe update the test instead of removing?

nice-pink replied:
I reverted my changes as I don't know what thoughts were behind forbidding ValueFrom.Parameter, so I added an additional case for ValueFrom.ConfigMapKeyRef. I also reverted the test and added a new one for the second case.

juliev0 commented:
So sorry, I'm going through old PRs and I see that unfortunately this one has been forgotten. I will try to review it soon.

Argo Workflows needs more people approving PRs unfortunately. If your company is by chance interested, I am gauging interest in this idea: https://github.com/argoproj/argo-workflows/issues/12229

juliev0 commented on workflow/validate/validate.go:
```diff
@@ -1424,9 +1424,11 @@ func validateDAGTaskArgumentDependency(arguments wfv1.Arguments, ancestry []stri
+			// Disallow ValueFrom.Parameter, but allow ValueFrom.ConfigMapKeyRef
+			if param.ValueFrom == nil || param.ValueFrom.Parameter != "" {
+				return errors.Errorf(errors.CodeBadRequest, "missing value for parameter '%s'", param.Name)
+			}
```
Nice fix. I see that `ValueFrom` has a lot of sub-parameter options. How did you decide which to allow and which not to?

nice-pink replied:
I just wanted to concentrate on the description in the ticket and allow as little as possible. Further, I wanted to align the behaviour of steps and dag for this very example.

Could be that other options might be valid here as well, though. That should be proposed and discussed in a follow up ticket.

nice-pink replied:
I just wanted to concentrate on the description in the ticket and allow as little as possible. Further, I wanted to align the behaviour of steps and dag for this very example.

Could be that other options might be valid here as well, though. That should be proposed and discussed in a follow up ticket.

juliev0 replied:
allow as little as possible? But I think it will allow all of the other members to be set, right? (just `ValueFrom.Parameter` is prevented)

nice-pink replied:
You are completely right.

juliev0 commented:
Looks good now. I saw in the original issue it says "throws the error when submitting in dag template but works for steps." Out of curiosity, is there any validation for ValueFrom members in steps?

nice-pink commented:
No, there is no test for steps. Should I add one for consistency? The only test related is for Inputs.Parameters (`paramWithValueFromConfigMapRef`) in `validate_tetst.go`.

juliev0 commented:
> No, there is no test for steps. Should I add one for consistency? The only test related is for Inputs.Parameters (`paramWithValueFromConfigMapRef`) in `validate_tetst.go`.

Thanks for offering that. In general, I'd say consistency is really important (plus consolidating the code into a single method). In this case, I'm concerned that if we apply your change, people will start reporting that certain ValueFrom members aren't working for steps, since before everything was allowed. One thought is: what if we took your previous iteration of code that only denied `ValueFrom.Parameter` (assuming you're fairly certain that that won't be evaluated correctly in either case - maybe it could be tested if not 100% sure?) and create a common function for that somewhere? Let me know what you think.

nice-pink commented:
I re-read the whole code related and checked why the `if` was added [Issue 7710](https://github.com/argoproj/argo-workflows/issues/7701). I think I found a better solution.

There is already the `validateArguments` method used for `steps` and `dag`. The method I updated was `validateDAGTaskArgumentDependency` which really should only validate if the dependencies exist before referencing them. I understand why the fix 7710 was added, but it makes all values defined by `valueFrom` invalid, even though only `valueFrom.Parameter` was intended to be considered invalid. Further, `valueFrom.Parameter` was only considered invalid for `dag`, even though it is invalid for `steps` also.

So I propse the current updates, so that `validateDAGTaskArgumentDependency` is used for dependency validation only. Further, I updated `validateArgumentsValues` to throw an error, if the values are not valid. 

To me it seems like only `ConfigMapKeyRef`, `Default` and likely `Expression` are valid options here.

Let me know what you think. If you also believe this makes sense, I'll add some more tests.

juliev0 commented:
Oh darn, I'm not sure if you saw, but it seems that your last commit accidentally altered a bunch of other files as well. 

As for your suggestion, it sounds great. I appreciate the digging you did. 

You mentioned: "To me it seems like only ConfigMapKeyRef, Default and likely Expression are valid options here." Would you mind going through each of the other members and commenting here on why each does not make sense? We should make sure that your assumption applies to all 3 places I see `validateArguments()` is called from. In general, I tend to think that validation is a "nice extra", and if we're unsure at all, then maybe less validation is better than more.

Really appreciate the iterations. Thank you!

nice-pink commented:
Thanks for noticing me about the messy commit!

I agree with less validation could be an option as currently the abilities are cut to prevent one invalid case.

Here are all options of `valueFrom`:

```
// Path in the container to retrieve an output parameter value from in container templates
Path string `json:"path,omitempty" protobuf:"bytes,1,opt,name=path"`
```
Is used for output parameters only.

```
// JSONPath of a resource to retrieve an output parameter value from in resource templates
JSONPath string `json:"jsonPath,omitempty" protobuf:"bytes,2,opt,name=jsonPath"`
```
Is used for output parameters only.

```
// JQFilter expression against the resource object in resource templates
JQFilter string `json:"jqFilter,omitempty" protobuf:"bytes,3,opt,name=jqFilter"`
```
Seems to be used for output parameters only.

```
// Selector (https://github.com/antonmedv/expr) that is evaluated against the event to get the value of the parameter. E.g. `payload.message`
Event string `json:"event,omitempty" protobuf:"bytes,7,opt,name=event"`
```
Actually not sure.

```
// Parameter reference to a step or dag task in which to retrieve an output parameter value from
// (e.g. '{{steps.mystep.outputs.myparam}}')
Parameter string `json:"parameter,omitempty" protobuf:"bytes,4,opt,name=parameter"`
```
Maybe parameter never makes sense, as parameters are referenced using `value: {{...}}`?!

```
// Supplied value to be filled in directly, either through the CLI, API, etc.
Supplied *SuppliedValueFrom `json:"supplied,omitempty" protobuf:"bytes,6,opt,name=supplied"`
```
Should only be used for `inputs.Parameters`.

```
// ConfigMapKeyRef is configmap selector for input parameter configuration
ConfigMapKeyRef *apiv1.ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,9,opt,name=configMapKeyRef"`
```
Valid

```
// Default specifies a value to be used if retrieving the value from the specified source fails
Default *AnyString `json:"default,omitempty" protobuf:"bytes,5,opt,name=default"`
```
Valid, e.g. if config map can't be read.

```
// Expression, if defined, is evaluated to specify the value for the parameter
Expression string `json:"expression,omitempty" protobuf:"bytes,8,rep,name=expression"`
```
Would be nice, to e.g. set predefined strings based on boolean set in `inputs.Parameters`.

juliev0 commented:
Thanks for walking through all of those. For a bunch you mentioned that they're only for Output Parameters. Is `validateArgumentsValues()` only being called for Input Parameters then? I'm not sure i see the code for that.

nice-pink commented:
Yes, these methods are only used for validating the arguments. The outputs are validated in `validateOutputs` and `validateOutputParameter`, the inputs are validated in `validateInputs`.

juliev0 commented:
> Yes, these methods are only used for validating the arguments. The outputs are validated in `validateOutputs` and `validateOutputParameter`, the inputs are validated in `validateInputs`.

Yes, thanks!

juliev0 commented:
Okay, so how about disallowing any that you're fairly certain should be and allowing the rest? (and maybe adding a "TODO" if not sure about some)

nice-pink commented:
Sounds good to me. I'll update the code and the test asap.

juliev0 commented on workflow/validate/validate.go:
```diff
@@ -856,11 +856,23 @@ func validateArgumentsFieldNames(prefix string, arguments wfv1.Arguments) error
+		// check for invalid valueFrom sub-parameters
+		if param.ValueFrom != nil && param.ValueFrom.Path != "" && param.ValueFrom.JSONPath != "" && param.ValueFrom.Event != "" && param.ValueFrom.Parameter != "" && param.ValueFrom.Supplied != nil {
+			return errors.Errorf(errors.CodeBadRequest, "only default, configMapKeyRef and supplied allowed for valueFrom '%s'", param.Name)
+		}
```
should those `&&` be `II` by chance? I thought if you simply set `param.ValueFrom.JSONPath` that it would be invalid.?

nice-pink replied:
You are completely right.

juliev0 commented on workflow/validate/validate.go:
```diff
@@ -856,11 +856,23 @@ func validateArgumentsFieldNames(prefix string, arguments wfv1.Arguments) error
+		// check for invalid valueFrom sub-parameters
+		if param.ValueFrom != nil && (param.ValueFrom.Path != "" || param.ValueFrom.JSONPath != "" || param.ValueFrom.Event != "" || param.ValueFrom.Parameter != "" || param.ValueFrom.Supplied != nil) {
+			return errors.Errorf(errors.CodeBadRequest, "only default, configMapKeyRef and supplied allowed for valueFrom '%s'", param.Name)
+		}
```
you mentioned in your notes that you weren't sure about `Event` - are you now?

juliev0 replied:
I think if I search for `Event`, it is being used as an argument, no? (in `server/event/dispatch/operation.go`)

nice-pink replied:
I haven't found this one. Thanks for checking.

juliev0 commented on workflow/validate/validate.go:
```diff
@@ -856,11 +856,23 @@ func validateArgumentsFieldNames(prefix string, arguments wfv1.Arguments) error
+		// check for invalid valueFrom sub-parameters
+		if param.ValueFrom != nil && (param.ValueFrom.Path != "" || param.ValueFrom.JSONPath != "" || param.ValueFrom.Event != "" || param.ValueFrom.Parameter != "" || param.ValueFrom.Supplied != nil) {
+			return errors.Errorf(errors.CodeBadRequest, "only default, configMapKeyRef and supplied allowed for valueFrom '%s'", param.Name)
+		}
```
This is a total nit, but can you put the whole thing inside one block so it's together for readability?:
```
if param.ValueFrom != nil {

}
```

juliev0 commented on workflow/validate/validate.go:
```diff
@@ -856,11 +856,25 @@ func validateArgumentsFieldNames(prefix string, arguments wfv1.Arguments) error
 				return errors.Errorf(errors.CodeBadRequest, "%s%s.value is required", prefix, param.Name)
 			}
 		}
+		// check for valid valueFrom sub-parameters
```
I think this comment needs to be moved to the first block below?

juliev0 commented on workflow/validate/validate.go:
```diff
@@ -856,11 +856,25 @@ func validateArgumentsFieldNames(prefix string, arguments wfv1.Arguments) error
+			// TODO: check if ValueFrom.Expression is valid!
+			if param.ValueFrom.ConfigMapKeyRef == nil && param.ValueFrom.Expression == "" {
+				return errors.Errorf(errors.CodeBadRequest, "only default, configMapKeyRef and supplied allowed for valueFrom '%s'", param.Name)
+			}
```
is it allowed to have `Event` and not `ConfigMapKeyRef` or `Expression`?

juliev0 replied:
I think from what I'm seeing, `Event` should be allowed but not `Expression`. 

Which IDE do you use? You should be able to do "Find all references" or the equivalent for each member of `ValueFrom` struct. That's where I can see who's calling it and how it's being used. It would be great if you might also do this to double check my findings.

nice-pink replied:
Sorry for the very late response.

You are completely right, that is the current state. Even though expression is currently not used for input/args I already had cases where I would love to use it.

Imagine you have a workflow which has "isDev" bool as input. A sub workflow now needs to set an environment dev/prod/... based on the isDev bool. Currently we need to define the sub workflow twice, set the environment parameter of each explicitly and decide on which sub workflow to trigger based on a `when: "isDev == true"` condition. In these (rare) cases it would be nice to have expression as an option for inputs. It would then be used in arguments.

Anyways, this is not the case currently.

shaharb-hs commented:
Any ETA on this? 🙏

nice-pink commented:
Sorry, I was pretty busy lately. I hope to complete the work on this PR this week.

juliev0 commented:
Is this ready for more review?

nice-pink commented:
Mabe thumbs up was not sufficient as reaction. Yes the PR is ready for review.

juliev0 commented:
> Mabe thumbs up was not sufficient as reaction. Yes, the PR is ready for review.

Okay, will look when I have time

juliev0 reviewed (changes requested):
I approved before checking the CI. Can you check the unit test failure? It may be related.

nice-pink commented:
I don't see a relation between my changes and the failing e2e tests.

juliev0 commented:
Okay, yes, those sometimes fail intermittently. Sorry for the hassle but can you submit [empty commits](https://www.freecodecamp.org/news/how-to-push-an-empty-commit-with-git/) until everything passes and then let me know?
